Make model exporting asynchronous

cdwcgt opened the pull request:
close #18476 

~~At the beginning, it was written according to `LegacyImportManager`.~~
~~so I create a `LegacyExportManager`, but seen unnecessary.~~

~~But doing this prevents getting dependencies again on every component that needs to be used, might be useful?~~

removed

https://user-images.githubusercontent.com/34775378/202493609-59f5a2b0-fe42-4f6b-b3e7-62663bd128cf.mp4

cdwcgt committed:
create a task to export to avoid block main thread

Code quality and remove some #nullable disable

cdwcgt committed:
add `LegacyExportManager`

cdwcgt committed:
working with test

cdwcgt committed:
impossible null and remove storage

cdwcgt committed:
Add comment

cdwcgt committed:
Make notification cannot cancel when Saving Zip Archive

This operation cannot be stopped(if not dispose stream).
so make it cannot cancel

bdach commented on osu.Game/Database/RealmAccess.cs:
```diff
@@ -173,11 +173,6 @@ public RealmAccess(Storage storage, string filename, GameThread? updateThread =
             if (!Filename.EndsWith(realm_extension, StringComparison.Ordinal))
                 Filename += realm_extension;
 
-#if DEBUG
-            if (!DebugUtils.IsNUnitRunning)
-                applyFilenameSchemaSuffix(ref Filename);
-#endif
```
What is the purpose of this change?

cdwcgt replied:
This should be a wrong change. Maybe I accidentally commited it during the test

bdach commented on osu.Game/Database/LegacyExportManager.cs:
```diff
@@ -0,0 +1,65 @@
+
+            switch (item)
+            {
+                case SkinInfo:
+                    await new LegacySkinExporter(exportStorage, realmAccess, notification, stream).ExportASync(item);
+                    break;
+
+                case ScoreInfo:
+                    await new LegacyScoreExporter(exportStorage, realmAccess, notification, stream).ExportASync(item);
+                    break;
+
+                case BeatmapSetInfo:
+                    await new LegacyBeatmapExporter(exportStorage, realmAccess, notification, stream).ExportASync(item);
+                    break;
```
Not sure I see much point to this class.

cdwcgt commented on osu.Game/Database/LegacyModelExporter.cs:
```diff
@@ -0,0 +1,127 @@
+        /// </summary>
+        protected abstract string FileExtension { get; }
+
+        protected readonly Storage UserFileStorage;
+
+        protected readonly Storage ExportStorage;
+
+        protected readonly RealmAccess RealmAccess;
+
+        protected readonly ProgressNotification Notification;
+
+        protected string Filename = null!;
+
+        protected Stream? OutputStream;
```
At first, I was set to a private class, but when rewriting the method, I found that these attributes were needed, so I changed to protected.

The rewriting method should be reconsidered..

cdwcgt committed:
Recover accidentally deleted codes

cdwcgt committed:
typo fix

cdwcgt committed:
refactor based on reviews

removed LegacyExportManager

Separated the method of CreateZip method and the default export method

cdwcgt committed:
log the error

cdwcgt committed:
add xmldoc

cdwcgt committed:
filename can be private

cdwcgt committed:
impossible null

bdach commented:
@peppy before this goes too far, curious of your thoughts on how this should be structured. I gave a rough sketch of how I'd see this parceled in the review above but I'm not super sure if I'm not overengineering this. If you're fine with the local resolves peppered around then maybe this is fine as a start.

peppy commented on osu.Game/Database/LegacyModelExporter.cs:
```diff
@@ -0,0 +1,137 @@
+                Text = "Exporting...",
+                CompletionText = "Export completed"
+            };
+            notification.CancelRequested += () => canCancel;
```
How does this even work? It's checking whether it.. can cancel but not actually performing any cancellation?

cdwcgt replied:
I use `notification.CancellationToken.IsCancellationRequested` to confirm whether task is request cancel.

when export zip stream, the task cannot be cancelled, so I use `canCancel` to avoid user cancel the task

peppy commented on osu.Game/Database/LegacyModelExporter.cs:
```diff
@@ -0,0 +1,137 @@
+        /// <summary>
+        /// Export te model corresponding to uuid to given stream.
+        /// </summary>
+        /// <param name="uuid">The medel which have <see cref="IHasGuidPrimaryKey"/>.</param>
```
The model to export?

bdach replied:
FWIW I basically ignored all xmldocs when reviewing this and will probably rewrite them in a final pass. Basically only wanted to ask about structure for now.

peppy commented on osu.Game/Database/LegacyModelExporter.cs:
```diff
@@ -0,0 +1,137 @@
+        /// <param name="uuid">The medel which have <see cref="IHasGuidPrimaryKey"/>.</param>
+        /// <param name="stream">The stream to export.</param>
+        /// <returns></returns>
+        public virtual async Task ExportToStreamAsync(TModel uuid, Stream stream)
```
```suggestion
        public virtual async Task ExportToStreamAsync(TModel model, Stream stream)
```

peppy commented on osu.Game/Database/LegacyModelExporter.cs:
```diff
@@ -0,0 +1,137 @@
+            }).ContinueWith(OnComplete);
+        }
+
+        private void createZipArchive(TModel model, Stream outputStream)
```
Seems about correct.

bdach replied:
@cdwcgt still unaddressed

cdwcgt replied:
I think its should already use this structure?
![image](https://user-images.githubusercontent.com/34775378/219875824-c637e2bd-bf98-4e88-b71e-24a27d80ccdd.png)
I don't know what different, please tell me pls

bdach replied:
I missed that because there are multiple classes in one file. Please split the files, one class per one file.

cdwcgt replied:
ok

peppy commented on osu.Game/Online/Leaderboards/LeaderboardScore.cs:
```diff
@@ -66,15 +67,21 @@ public class LeaderboardScore : OsuClickableContainer, IHasContextMenu, IHasCust
         [Resolved]
         private Storage storage { get; set; }
 
+        [Resolved]
+        private RealmAccess realm { get; set; }
+
+        [Resolved(canBeNull: true)]
+        private INotificationOverlay notifications { get; set; }
```
I tend to agree that making part of the `Manager` classes is best. Going forward, we'll probably have an `ExportLegacy` and an `Export` (for the newer format type, if/when that exists) and it makes sense to keep them both in the same place.

Put another way, this isn't doing anything *so legacy* that it needs to be separate code-side (unlike some of the other stuff which is in the legacy namespace).

cdwcgt committed:
Spite sync method `ExportToStream`

`uuid` to `model`

cdwcgt commented:
So should we use DI? Or a manager class or keep.
I make a solution that uses DI, but it feels less than ideal in terms of code quality I think?

cdwcgt committed:
Merge branch 'master' into export

cdwcgt committed:
Fix overwriting existing files

https://github.com/ppy/osu/pull/21468

cdwcgt committed:
remove Component

exportStorage to private
notification should post when export instead of class being constructed

cdwcgt committed:
ExportAsync use TModel

cdwcgt committed:
`onComplete` should private

cdwcgt committed:
rename method name and add xmldoc

cdwcgt committed:
xmldoc

cdwcgt committed:
Group export methods into their respective managers

bdach commented on osu.Game/Beatmaps/BeatmapManager.cs:
```diff
@@ -446,6 +453,8 @@ public void UndeleteAll()
         public Task<Live<BeatmapSetInfo>?> ImportAsUpdate(ProgressNotification notification, ImportTask importTask, BeatmapSetInfo original) =>
             beatmapImporter.ImportAsUpdate(notification, importTask, original);
 
+        public void Export(BeatmapSetInfo beatmap) => Task.Run(() => beatmapExporter.ExportAsync(beatmap));
```
The threading here is making zero sense. The `ExportAsync()` call (which is *async*) is being wrapped in a `Task.Run()`? And then there is another `Task.Run()` *inside* of `ExportAsync()`? So two TPL threads are now involved? What is going on here?

This should most likely not use `Task.Run()`, and not return `void`. Instead it should return `Task`, and usage sites that don't support async well should use the `.FireAndForget()` extension to get at least a semblance of proper error checking.

cdwcgt replied:
I'm dizzy that time, sorry

bdach commented on osu.Game/Online/Leaderboards/LeaderboardScore.cs:
```diff
@@ -96,6 +93,8 @@ private void load(IAPIProvider api, OsuColour colour, ScoreManager scoreManager)
 
             statisticsLabels = GetStatistics(Score).Select(s => new ScoreComponentLabel(s)).ToList();
 
+            this.scoreManager = scoreManager;
```
If you're going to do this then you may as well move `ScoreManager` out from the BDL param to a `[Resolved]` field.

bdach commented on osu.Game/Skinning/SkinManager.cs:
```diff
@@ -280,6 +287,8 @@ public Action<IEnumerable<Live<SkinInfo>>> PresentImport
 
         public Task<Live<SkinInfo>> Import(ImportTask task, bool batchImport = false, CancellationToken cancellationToken = default) => skinImporter.Import(task, batchImport, cancellationToken);
 
+        public void ExporCurrenttSkin() => CurrentSkinInfo.Value.PerformRead(s => skinExporter.ExportAsync(s));
```
Typo in method name.

Also method should arguably not export "current skin", it should export a skin as given by a parameter.

cdwcgt committed:
fix weird async logic

cdwcgt committed:
use resolved attribute

cdwcgt committed:
change abstract implement

cdwcgt committed:
Export instead of ExportCurrentSkin

cdwcgt committed:
use ThrowIfCancellationRequested instead of CancelRequested

cdwcgt committed:
use ZipWriter

Export directly to stream instead of creating a archive
so we can cancel this anytime

cdwcgt committed:
remove try catch

cdwcgt committed:
Handle the case where the file cannot be found

cdwcgt committed:
using declaration

reshaper

cdwcgt committed:
cancellationToken pass, notification adujust

cdwcgt committed:
Merge branch 'master' into export

cdwcgt committed:
ConfigureAwait for awaited task

cdwcgt committed:
GetFilename and something other

https://github.com/ppy/osu/pull/21739

cdwcgt committed:
split LegacyArchiveExporter

cdwcgt committed:
catch OperationCanceledException

cdwcgt committed:
catch ObjectDisposedException

cdwcgt committed:
store exportingModels for all exporter

No one wants to export multiple copies of the same model at the same time, right?

cdwcgt committed:
comment

cdwcgt committed:
catch when zipWriter dispose

ObjectDisposedException also appear when zipwriter dispose after user request cancel

cdwcgt committed:
Accidentally deleted using

wtf

cdwcgt committed:
remove ()

cdwcgt committed:
Path.GetExtension() will not get null

bdach commented on osu.Game/Database/LegacyArchiveExporter.cs:
```diff
@@ -0,0 +1,85 @@
+
+                    using (var stream = UserFileStorage.GetStream(file.File.GetStoragePath()))
+                    {
+                        // Sometimes we cannot find the file(probably deleted by the user), so we handle this and post a error.
+                        if (stream == null)
+                        {
+                            // Only pop up once to prevent spam.
+                            if (!fileMissing)
+                            {
+                                PostNotification?.Invoke(new SimpleErrorNotification
+                                {
+                                    Text = "Some of your files are missing, they will not be included in the archive"
```
What on earth? This is shocking. What is this logic even? Why is it here? And when is it ever tolerable to tell a user that "some of their files are missing"?

cdwcgt replied:
This was an accidental case for me, maybe because I imported a skin that used the same file and deleted one, or some other accident, the realm couldn't find the skin file.
If this exception occurs, the entire package will not be exported in old logic, but this is not fatal, so I choose to pop up the prompt and continue

peppy replied:
I don't yet know how we should handle this kind of thing. I can imagine a scenario where this does happen, and I agree we probably want to present this to the user in some way, but I'm not 100% sure doing it with a notification like this is correct.

Similar cases could come up, for instance, when starting gameplay and files are missing. I'd hope we could standardise the way these are presented across the game. So probably wouldn't be against removing this localised handling for now (or at least replacing with a loud `Logger.Log`).

bdach commented on osu.Game/Database/LegacyArchiveExporter.cs:
```diff
@@ -0,0 +1,85 @@
+                    notification.Text = $"Exporting... ({i}/{model.Files.Count()})";
+                }
+            }
+            catch (ObjectDisposedException)
+            {
+                // outputStream may close before writing when request cancel.
+                if (cancellationToken.IsCancellationRequested)
+                    return;
+
+                throw;
+            }
```
I don't even know what is happening here.

cdwcgt replied:
It looks like if the user cancel
https://github.com/ppy/osu/pull/21308/files#diff-6a0e3884e01dc22403790923a038a2cdda6d02f121b621682c78bc3b12b7c833R100
it will run out of range first and dispose the stream and cause `ObjectDisposedExcption`
Because it can easily happen during the cancellation process, I choose to catch and only return when the user cancels

peppy commented on osu.Game/Database/LegacyModelExporter.cs:
```diff
@@ -0,0 +1,164 @@
+            }
+            finally
+            {
+                // cleanup if export is failed or canceled.
+                if (!success)
+                {
+                    exportStorage.Delete(filename);
+                }
```
Why not move this outside of the finally and `return` in the `catch`?

cdwcgt replied:
If it is not finally, the exported error (An error other than `OperationCanceledException` has occurred) file should not be deleted.
If it is determined that keeping the wrong file is the expected behavior, then I will change it,

peppy commented on osu.Game/Database/LegacyModelExporter.cs:
```diff
@@ -0,0 +1,164 @@
+                    success = await ExportToStreamAsync(model, stream, notification, cancellationToken ?? notification.CancellationToken).ConfigureAwait(false);
+                }
+            }
+            catch (OperationCanceledException)
```
We probably want to look into the best practices of catching `OperationCanceledException`. I'm pretty sure we don't want to catch without a `rethrow`, but we're already doing this (likely incorrectly) in other places.

peppy commented on osu.Game/Database/LegacyModelExporter.cs:
```diff
@@ -0,0 +1,164 @@
+            ProgressNotification notify = notification ?? new ProgressNotification();
+
+            Guid id = model.ID;
+            return await Task.Run(() =>
```
I'm confused. Why are you using `Task.Run` here? You're in an `async` method. Recommendation would be

```diff
diff --git a/osu.Game/Database/LegacyModelExporter.cs b/osu.Game/Database/LegacyModelExporter.cs
index acb83ce1d5..03f4c06297 100644
--- a/osu.Game/Database/LegacyModelExporter.cs
+++ b/osu.Game/Database/LegacyModelExporter.cs
@@ -124,12 +124,13 @@ public async Task ExportAsync(TModel model, CancellationToken? cancellationToken
         /// <param name="notification">The notification will displayed to the user</param>
         /// <param name="cancellationToken">The Cancellation token that can cancel the exporting.</param>
         /// <returns>Whether the export was successful</returns>
-        public async Task<bool> ExportToStreamAsync(TModel model, Stream stream, ProgressNotification? notification = null, CancellationToken cancellationToken = default)
+        public Task<bool> ExportToStreamAsync(TModel model, Stream stream, ProgressNotification? notification = null, CancellationToken cancellationToken = default)
         {
             ProgressNotification notify = notification ?? new ProgressNotification();
 
             Guid id = model.ID;
-            return await Task.Run(() =>
+
+            return Task.Run(() =>
             {
                 realmAccess.Run(r =>
                 {
@@ -148,7 +149,7 @@ public async Task<bool> ExportToStreamAsync(TModel model, Stream stream, Progres
                 notify.CompletionText = "Export Complete, Click to open the folder";
                 notify.State = ProgressNotificationState.Completed;
                 return true;
-            }, cancellationToken).ConfigureAwait(false);
+            }, cancellationToken);
         }
 
         /// <summary>

```

peppy commented on osu.Game/Database/LegacyModelExporter.cs:
```diff
@@ -0,0 +1,164 @@
+                });
+            }, cancellationToken).ContinueWith(t =>
+            {
+                if (t.IsFaulted)
```
No logic to handle cancellation?

cdwcgt replied:
because of `cancellationToken`, this task will be canceled
I have stayed this before:
```C#
if (t.IsCanceled)
   return;
```

peppy commented on osu.Game/Database/LegacyModelExporter.cs:
```diff
@@ -0,0 +1,164 @@
+
+        /// <summary>
+        /// Exports an item to Stream.
+        /// Override if custom export method is required.
```
Doesn't make sense.

cdwcgt replied:
The previous compression logic is here (virtual method), forgot to change the comment

peppy commented on osu.Game/Database/LegacyModelExporter.cs:
```diff
@@ -0,0 +1,164 @@
+            {
+                State = ProgressNotificationState.Active,
+                Text = "Exporting...",
+                CompletionText = "Export completed"
```
Just remove this?

cdwcgt replied:
For `CompletionText`, I know it will be rewritten later, but I keep a generic text for other cases.
For example, the target platform does not support clicking to open the folder.

Maybe I'm overthinking it, but let me know if it's not necessary and I'll change it

peppy commented on osu.Game/Database/LegacyModelExporter.cs:
```diff
@@ -0,0 +1,164 @@
+            {
+                PostNotification?.Invoke(new SimpleErrorNotification
+                {
+                    Text = "File is being exported"
```
???

peppy replied:
Don't do this. Remove this weird handling.

peppy commented on osu.Game/Database/LegacyModelExporter.cs:
```diff
@@ -0,0 +1,164 @@
+                    return false;
+                }
+
+                notify.CompletionText = "Export Complete, Click to open the folder";
```
Incorrect grammar. Also it doesn't really open the folder but present the file.

Maybe look at https://github.com/ppy/osu/blob/f8537c1cbe98bc38f0423e8eabdc2332a21d22ab/osu.Game/Database/RealmArchiveModelImporter.cs#L157-L172 for inspiration.

peppy reviewed (changes requested):
zzzz

cdwcgt committed:
async logic fix

cdwcgt committed:
wrong xmldoc

because of https://github.com/ppy/osu/pull/21308/commits/6900d0120a70287ef531a4c0facd398508a76b06

cdwcgt committed:
delete notify post when duplicate export

cdwcgt committed:
use log

cdwcgt committed:
make static

cdwcgt committed:
log to database

cdwcgt committed:
make static name fix

cdwcgt committed:
cancel handle

cdwcgt committed:
logic fix?

cdwcgt committed:
typo

cdwcgt committed:
remove `CompletionText` in Constructor

this will never used

Co-Authored-By: Dean Herbert <pe@ppy.sh>

cdwcgt committed:
Merge remote-tracking branch 'upstream/master' into export

cdwcgt committed:
exception handling

cdwcgt committed:
merge fix

wtf?

cdwcgt committed:
fix test in #22653

cdwcgt committed:
code quality

cdwcgt committed:
repair usage of CancellationToken

Co-Authored-By: n0099 <n@n0099.net>

cdwcgt committed:
notification nullable fix

cdwcgt committed:
use Live<TModel>

Use RealmAccess only when needed

cdwcgt committed:
CompletionText

Co-Authored-By: Dean Herbert <pe@ppy.sh>

cdwcgt committed:
Merge branch 'master' into export

peppy committed:
Rename class back to original name so I can read the diff

peppy commented on osu.Game/Database/LegacyModelExporter.cs:
```diff
@@ -0,0 +1,196 @@
+    /// <summary>
+    /// A class which handles exporting legacy user data of a single type from osu-stable.
+    /// </summary>
+    public abstract class LegacyModelExporter<TModel>
```
I'm renaming this back because the way you made changes and committed them makes it impossible to know what (if anything )in this class was changed.

peppy committed:
Remove need for `TestRealmLive` nonsense

peppy committed:
Remove unnecessary local variable

peppy committed:
Improve xmldoc and fix silly progress text in `LegacyArchiveExporter`

peppy committed:
Move file missing logging to a better location, and log actual filenames to logs

peppy committed:
Fix notification text not including export filename

peppy committed:
Simplify `ExportToStreamAsync` implementation

peppy committed:
Simplify `ExportAsync`, remove weird dedupe logic and unnecessary return `success` code

peppy committed:
Remove weird catch logic

peppy committed:
Fix incorrect cancellation logic

peppy commented:
I've basically rewritten this PR.

@bdach requesting your fresh eyes on it.
@cdwcgt please, read through each and every one of my commits multiple times and understand why the changes I made were made. ask if you have any questions.

I've left the rename in so the diff is legible (https://github.com/ppy/osu/pull/21308/commits/01581024678c6f4d99a54892cb95cbe7599f0f3c). It can be renamed back before merge.

peppy commented on osu.Game/Database/LegacyExporter.cs:
```diff
@@ -38,56 +39,107 @@ public abstract class LegacyExporter<TModel>
+
+        public Action<Notification>? PostNotification { get; set; }
 
+        /// <summary>
+        /// Construct exporter.
+        /// Create a new exporter for each export, otherwise it will cause confusing notifications.
+        /// </summary>
+        /// <param name="storage">Storage for storing exported files. Basically it is used to provide export stream</param>
```
xmldoc still needs rewriting. i'll get to it later if i'm not beaten.

peppy commented on osu.Game/Database/LegacyExporter.cs:
```diff
@@ -38,56 +39,107 @@ public abstract class LegacyExporter<TModel>
+        /// <param name="outputStream">The output stream to export to.</param>
+        /// <param name="notification">The notification will displayed to the user</param>
+        /// <param name="cancellationToken">The Cancellation token that can cancel the exporting.</param>
+        protected abstract void ExportToStream(TModel model, Stream outputStream, ProgressNotification? notification, CancellationToken cancellationToken = default);
```
Probably no reason to be `protected`.

cdwcgt commented:
https://github.com/ppy/osu/pull/21308/commits/2a3e03695cad820002999e420797de7a0faf4cba#diff-d5a620a82a32a4027cf0f259048441c48a70c4825f93636aa06a4c4aefb102caL48-L50
Shouldn't we prevent duplicate exports, like clicking twice "export the current skin"

peppy commented:
I don't see any reason to prevent it.

cdwcgt committed:
ExportToStream should be public

cdwcgt committed:
remove meanless comment

cdwcgt commented:
ok then I should be fine.
I will change `ExportToStream` change to public, and remove [this comment](https://github.com/ppy/osu/pull/21308/files#diff-79d1a2d699ac04e9d6b897b9116fbae2c542fadb0b1ba6c86f58995d467d15ceR72)

it should be ok after changing the class name, if no other problem.

peppy committed:
Remove redundant xmldoc and reword some remaining

peppy committed:
Remove non-`Live` pathway

peppy committed:
More xmldoc fixes

peppy committed:
Rename `outputStream` variable to match base class

bdach committed:
Make `UserFileStorage` `readonly` again

bdach committed:
Move lower and xmldoc `GetFilename(TModel)`

bdach committed:
Remove weird single-use private method

bdach committed:
Unify error message wording

bdach committed:
Rename flag

bdach committed:
Reorder increment to make progress update less weird

bdach committed:
Mark exporter extensions as untranslatable

bdach commented on osu.Game/Database/LegacyScoreExporter.cs:
```diff
@@ -28,7 +26,9 @@ protected override string GetFilename(ScoreInfo score)
-        public override void ExportModelTo(ScoreInfo model, Stream outputStream)
+        protected override string FileExtension => ".osr";
+
+        public override void ExportToStream(ScoreInfo model, Stream outputStream, ProgressNotification? notification, CancellationToken cancellationToken = default)
```
This isn't handling the case of missing file in the same way as `LegacyArchiveExporter` - but I'm looking away not to add more stuff onto of this trash fire of a branch than absolutely necessary (and I still want to push in a few *other* touch-ups).

bdach committed:
Merge branch 'master' into export
